io/s3.go

```diff
@@ -0,0 +1,108 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```

Comment:
overall this file LGTM. what would be great is to have some tests that use S3 against a minio container, which can be done in a follow-up PR